pkg/agent/packetcapture/packetcapture_controller.go

```diff
@@ -0,0 +1,774 @@
+	device := c.getTargetCaptureDevice(pc)
+	klog.V(2).InfoS("Prepare capture on current node", "name", pc.Name, "device", *device)
+	pcState.targetCapturedPacketsNum = pc.Spec.CaptureConfig.FirstN.Number
+	file, writer, err := getPacketFileAndWriter(pc.Name)
+	if err != nil {
+		return err
+	}
+	pcState.pcapngFile = file
+	pcState.pcapngWriter = writer
+	pcState.updateRateLimiter = rate.NewLimiter(rate.Every(captureStatusUpdatePeriod), 1)
+	timeout := defaultTimeoutDuration
+	if pc.Spec.Timeout != nil {
+		timeout = time.Duration(*pc.Spec.Timeout) * time.Second
+	}
+	go func() {
+		captureErr := c.performCapture(pc, pcState, *device, srcIP, dstIp, timeout)
+		func() {
+			c.lock.Lock()
+			defer c.lock.Unlock()
+			c.numRunningCaptures -= 1
+			c.captures[name] = packetCapturePhaseCompleted
+		}()
+		updateErr := c.updatePacketCaptureStatus(name, pcState.capturedPacketsNum, "", captureErr)
+		if updateErr != nil {
+			klog.ErrorS(updateErr, "Failed to update PacketCapture status")
+		}
+	}()
+	return nil
+}
+
+func (c *Controller) performCapture(
+	pc *crdv1alpha1.PacketCapture,
+	captureState *packetCaptureState,
+	device string,
+	srcIP, dstIP net.IP,
+	timeout time.Duration,
+) error {
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+	packets, err := c.CaptureInterface.Capture(ctx, device, srcIP, dstIP, pc.Spec.Packet)
+	if err != nil {
+		klog.ErrorS(err, "Failed to start capture")
+		return err
+	}
+
+	for {
+		select {
+		case packet := <-packets:
+			if captureState.capturedPacketsNum == captureState.targetCapturedPacketsNum {
+				return nil
+			}
+			captureState.capturedPacketsNum++
+			ci := gopacket.CaptureInfo{
+				Timestamp:     time.Now(),
+				CaptureLength: len(packet.Data()),
+				Length:        len(packet.Data()),
+			}
+			err = captureState.pcapngWriter.WritePacket(ci, packet.Data())
+			if err != nil {
+				return fmt.Errorf("couldn't write packets: %w", err)
+			}
+			klog.V(5).InfoS("Capture packets", "name", captureState.name, "count",
+				captureState.capturedPacketsNum, "len", ci.Length)
+
+			reachTarget := captureState.capturedPacketsNum == captureState.targetCapturedPacketsNum
+			// use rate limiter to reduce the times we need to update status.
+			if reachTarget || captureState.updateRateLimiter.Allow() {
+				pc, err := c.packetCaptureLister.Get(captureState.name)
+				if err != nil {
+					return fmt.Errorf("get PacketCapture failed: %w", err)
+				}
+				// if reach the target. flush the file and upload it.
+				if reachTarget {
+					path := os.Getenv("POD_NAME") + ":" + nameToPath(pc.Name)
+					statusPath := path
+					if err = captureState.pcapngWriter.Flush(); err != nil {
+						return err
+					}
+					if pc.Spec.FileServer != nil {
+						err = c.uploadPackets(pc, captureState.pcapngFile)
+						klog.V(4).InfoS("Upload captured packets", "name", pc.Name, "path", path)
+						statusPath = fmt.Sprintf("%s/%s.pcapng", pc.Spec.FileServer.URL, pc.Name)
```

Comment:
I feel we should check the above err and set statusPath as empty string if it's uploaded successfully.